README.md
# http-utils

[![deno land](http://img.shields.io/badge/available%20on-deno.land/x-lightgrey.svg?logo=deno)](https://deno.land/x/http_utils)
[![GitHub release (latest by date)](https://img.shields.io/github/v/release/httpland/http-utils)](https://github.com/httpland/http-utils/releases)
[![codecov](https://codecov.io/github/httpland/http-utils/branch/main/graph/badge.svg)](https://codecov.io/gh/httpland/http-utils)
[![GitHub](https://img.shields.io/github/license/httpland/http-utils)](https://github.com/httpland/http-utils/blob/main/LICENSE)

[![test](https://github.com/httpland/http-utils/actions/workflows/test.yaml/badge.svg)](https://github.com/httpland/http-utils/actions/workflows/test.yaml)
[![NPM](https://nodei.co/npm/@httpland/http-utils.png?mini=true)](https://nodei.co/npm/@httpland/http-utils/)

HTTP utility collection.

Compliant with [RFC 9110](https://www.rfc-editor.org/rfc/rfc9110).

## Header

Utilities for HTTP header field.

### isMessageMetadataHeader

Whether the input is [MessageMetadataHeader](#messagemetadataheader) or not.

```ts
import { isMessageMetadataHeader } from "https://deno.land/x/http_utils@$VERSION/header.ts";
import { assert } from "https://deno.land/std/testing/asserts.ts";

assert(isMessageMetadataHeader("date"));
assert(!isMessageMetadataHeader("<others>"));
```

### isMessageForwardingHeader

Whether the input is [MessageForwardingHeader](#messageforwardingheader) or not.

```ts
import { isMessageForwardingHeader } from "https://deno.land/x/http_utils@$VERSION/header.ts";
import { assert } from "https://deno.land/std/testing/asserts.ts";

assert(isMessageForwardingHeader("connection"));
assert(!isMessageForwardingHeader("<others>"));
```

### isRepresentationHeader

Whether the input is [RepresentationHeader](#representationheader) or not.

```ts
import { isRepresentationHeader } from "https://deno.land/x/http_utils@$VERSION/header.ts";
import { assert } from "https://deno.land/std/testing/asserts.ts";

assert(isRepresentationHeader("content-type"));
assert(!isRepresentationHeader("<others>"));
```

### isAuthenticationHeader

Whether the input is [AuthenticationHeader](#authenticationheader) or not.

```ts
import { isAuthenticationHeader } from "https://deno.land/x/http_utils@$VERSION/header.ts";
import { assert } from "https://deno.land/std/testing/asserts.ts";

assert(isAuthenticationHeader("authorization"));
assert(!isAuthenticationHeader("<others>"));
```

### isContentNegotiationHeader

Whether the input is [ContentNegotiationHeader](#contentnegotiationheader) or
not.

```ts
import { isContentNegotiationHeader } from "https://deno.land/x/http_utils@$VERSION/header.ts";
import { assert } from "https://deno.land/std/testing/asserts.ts";

assert(isContentNegotiationHeader("accept"));
assert(!isContentNegotiationHeader("<others>"));
```

### isConditionalHeader

Whether the input is [ConditionalHeader](#conditionalheader) or not.

```ts
import { isConditionalHeader } from "https://deno.land/x/http_utils@$VERSION/header.ts";
import { assert } from "https://deno.land/std/testing/asserts.ts";

assert(isConditionalHeader("if-match"));
assert(!isConditionalHeader("<others>"));
```

### isRangeHeader

Whether the input is [RangeHeader](#rangeheader) or not.

```ts
import { isRangeHeader } from "https://deno.land/x/http_utils@$VERSION/header.ts";
import { assert } from "https://deno.land/std/testing/asserts.ts";

assert(isRangeHeader("range"));
assert(!isRangeHeader("<others>"));
```

### isCachingHeader

Whether the input is [CachingHeader](#cachingheader) or not.

```ts
import { isCachingHeader } from "https://deno.land/x/http_utils@$VERSION/header.ts";
import { assert } from "https://deno.land/std/testing/asserts.ts";

assert(isCachingHeader("age"));
assert(!isCachingHeader("<others>"));
```

### MessageMetadataHeader

HTTP Message Metadata header fields.

Compliant with
[RFC 9110, 6.6. Message Metadata](https://www.rfc-editor.org/rfc/rfc9110.html#section-6.6).

- Date
- Trailer

```ts
import { MessageMetadataHeader } from "https://deno.land/x/http_utils@$VERSION/header.ts";
import { assertEquals } from "https://deno.land/std/testing/asserts.ts";

assertEquals(MessageMetadataHeader.Date, "date");
```

### MessageForwardingHeader

HTTP Message Forwarding header fields.

Compliant with
[RFC 9110, 7.6. Message Forwarding](https://www.rfc-editor.org/rfc/rfc9110.html#section-7.6).

- Connection
- Max-Forwards
- Via

```ts
import { MessageForwardingHeader } from "https://deno.land/x/http_utils@$VERSION/header.ts";
import { assertEquals } from "https://deno.land/std/testing/asserts.ts";

assertEquals(MessageForwardingHeader.Via, "via");
```

### RepresentationHeader

HTTP representation data and metadata header fields.

Compliant with
[RFC 9110, 8. Representations](https://www.rfc-editor.org/rfc/rfc9110.html#section-8).

- Content-Type
- Content-Encoding
- Content-Language
- Content-Length
- Content-Location
- Last-Modified
- ETag

```ts
import { RepresentationHeader } from "https://deno.land/x/http_utils@$VERSION/header.ts";
import { assertEquals } from "https://deno.land/std/testing/asserts.ts";

assertEquals(RepresentationHeader.ContentType, "content-type");
```

### AuthenticationHeader

HTTP Authentication header fields.

Compliant with
[RFC 9110, 11. HTTP Authentication](https://www.rfc-editor.org/rfc/rfc9110#section-11).

- WWW-Authenticate
- Authorization
- Authentication-Info
- Proxy-Authenticate
- Proxy-Authorization
- Proxy-Authentication-Info

```ts
import { AuthenticationHeader } from "https://deno.land/x/http_utils@$VERSION/header.ts";
import { assertEquals } from "https://deno.land/std/testing/asserts.ts";

assertEquals(AuthenticationHeader.Authorization, "authorization");
```

### ContentNegotiationHeader

HTTP content negotiation header fields.

Compliant with
[RFC 9110, 12. Content Negotiation](https://www.rfc-editor.org/rfc/rfc9110#section-12).

- Accept
- Accept-Charset
- Accept-Encoding
- Accept-Language
- Vary

```ts
import { ContentNegotiationHeader } from "https://deno.land/x/http_utils@$VERSION/header.ts";
import { assertEquals } from "https://deno.land/std/testing/asserts.ts";

assertEquals(ContentNegotiationHeader.Accept, "accept");
```

### ConditionalHeader

HTTP conditional requests header fields.

Compliant with
[RFC 9110, 13. Conditional Requests](https://www.rfc-editor.org/rfc/rfc9110#section-13).

- If-Match
- If-None-Match
- If-Modified-Since
- If-Unmodified-Since
- If-Range

```ts
import { ConditionalHeader } from "https://deno.land/x/http_utils@$VERSION/header.ts";
import { assertEquals } from "https://deno.land/std/testing/asserts.ts";

assertEquals(ConditionalHeader.IfNoneMatch, "if-none-match");
```

### RangeHeader

HTTP range requests header fields.

Compliant with
[RFC 9110, 14. Range Requests](https://www.rfc-editor.org/rfc/rfc9110#section-14).

- Range
- Accept-Ranges
- Content-Range

```ts
import { RangeHeader } from "https://deno.land/x/http_utils@$VERSION/header.ts";
import { assertEquals } from "https://deno.land/std/testing/asserts.ts";

assertEquals(RangeHeader.Range, "range");
```

### CachingHeader

HTTP Caching header fields.

Compliant with [RFC 9111, HTTP Caching](https://www.rfc-editor.org/rfc/rfc9111).

- Age
- Cache-Control
- Expires

```ts
import { CachingHeader } from "https://deno.land/x/http_utils@$VERSION/header.ts";
import { assertEquals } from "https://deno.land/std/testing/asserts.ts";

assertEquals(CachingHeader.CacheControl, "cache-control");
```

## Method

Utilities for HTTP method.

### isSafeMethod

Whether the method is safe method or not.

Defined in
[RFC 9110, 9.2.1. Safe Methods](https://www.rfc-editor.org/rfc/rfc9110.html#name-safe-methods).

```ts
import { isSafeMethod } from "https://deno.land/x/http_utils@$VERSION/method.ts";
import { assert } from "https://deno.land/std/testing/asserts.ts";

assert(isSafeMethod("GET"));
assert(isSafeMethod("HEAD"));
assert(isSafeMethod("OPTIONS"));
assert(isSafeMethod("TRACE"));
```

### isIdempotentMethod

Whether the method is idempotent method or not.

Defined in
[RFC 9110, 9.2.2 Idempotent Methods](https://www.rfc-editor.org/rfc/rfc9110.html#name-idempotent-methods).

```ts
import { isIdempotentMethod } from "https://deno.land/x/http_utils@$VERSION/method.ts";
import { assert } from "https://deno.land/std/testing/asserts.ts";

assert(isIdempotentMethod("GET"));
assert(isIdempotentMethod("PUT"));
assert(isIdempotentMethod("DELETE"));
```

### isRetrieveMethod

Whether the method is retrieve method or not.

Retrieve method is following:

- GET
- HEAD

```ts
import { isRetrieveMethod } from "https://deno.land/x/http_utils@$VERSION/method.ts";
import { assert } from "https://deno.land/std/testing/asserts.ts";

assert(isRetrieveMethod("GET"));
assert(isRetrieveMethod("HEAD"));
assert(!isRetrieveMethod("POST"));
```

## Lists

Compliant with
[RFC 9110, 5.6.1. Lists](https://www.rfc-editor.org/rfc/rfc9110.html#section-5.6.1).

### parseListFields

Parse
[list-based fields](https://www.rfc-editor.org/rfc/rfc9110.html#section-5.5-7)
into array.

Strings enclosed in double quotes are safely handled.

```ts
import { parseListFields } from "https://deno.land/x/http_utils@$VERSION/list.ts";
import { assertEquals } from "https://deno.land/std/testing/asserts.ts";

assertEquals(parseListFields("foo , ,bar,charlie"), [
  "foo",
  "bar",
  "charlie",
]);
assertEquals(parseListFields(`"Sat, 04 May 1996", "Wed, 14 Sep 2005"`), [
  `"Sat, 04 May 1996"`,
  `"Wed, 14 Sep 2005"`,
]);
```

## Tokens

Compliant with
[RFC 9110, 5.6.2. Tokens](https://www.rfc-editor.org/rfc/rfc9110.html#section-5.6.2).

### isTchar

Whether the input is
[tchar](https://www.rfc-editor.org/rfc/rfc9110.html#section-5.6.2-2) or not.

```ts
import { isTchar } from "https://deno.land/x/http_utils@$VERSION/token.ts";
import { assert, assertFalse } from "https://deno.land/std/testing/asserts.ts";

assert(isTchar("!"));
assert(isTchar("a"));
assert(isTchar("Z"));
assertFalse(isTchar(""));
```

### isToken

Whether the input is
[token](https://www.rfc-editor.org/rfc/rfc9110.html#section-5.6.2-2) or not.

```ts
import { isToken } from "https://deno.land/x/http_utils@$VERSION/token.ts";
import { assert, assertFalse } from "https://deno.land/std/testing/asserts.ts";

assert(isToken("token"));
assert(isToken("*!~"));
assertFalse(isToken(""));
```

## Quoted Strings

Compliant with
[RFC 9110, 5.6.4. Quoted Strings](https://www.rfc-editor.org/rfc/rfc9110.html#section-5.6.4).

### isQdtext

Whether the input is
[qdtext](https://www.rfc-editor.org/rfc/rfc9110.html#section-5.6.4-2) or not.

```ts
import { isQdtext } from "https://deno.land/x/http_utils@$VERSION/quoted_string.ts";
import { assert, assertFalse } from "https://deno.land/std/testing/asserts.ts";

assert(isQdtext("\t"));
assert(isQdtext("\xFF"));
assertFalse(isQdtext(`"`));
```

### isQuotedPair

Whether the input is
[quoted-pair](https://www.rfc-editor.org/rfc/rfc9110.html#section-5.6.4-4) or
not.

```ts
import { isQuotedPair } from "https://deno.land/x/http_utils@$VERSION/quoted_string.ts";

CHANGELOG.md
# [2.0.0-beta.1](https://github.com/httpland/http-utils/compare/1.3.0...2.0.0-beta.1) (2023-05-12)


### Features

* remove all deprecated modules ([41837aa](https://github.com/httpland/http-utils/commit/41837aac040113f50ba74c10e88691832e8a6152))
* **status:** add `ErrorStatus` enum ([4019655](https://github.com/httpland/http-utils/commit/401965598f13fba4687fa32f0f522234f3a13df7))
* **status:** add HTTP status code enums ([1ab8b6e](https://github.com/httpland/http-utils/commit/1ab8b6edeff25802ef35092a58e6786b93cdb22f))
* **status:** add type guard for HTTP status ([3416952](https://github.com/httpland/http-utils/commit/3416952ae693e23bde7b3d67ca532fd38e4b19bc))
* **status:** rename enum keys ([e8447be](https://github.com/httpland/http-utils/commit/e8447be4440c83f377680de844c6477c6d457e24))


### BREAKING CHANGES

* remove deprecated modules

# [1.3.0](https://github.com/httpland/http-utils/compare/1.2.0...1.3.0) (2023-05-09)


### Features

* mark as deprecated ([ab24b7d](https://github.com/httpland/http-utils/commit/ab24b7d2177aba3979a802fa02707a4a2a4ede9b))

# [1.3.0-beta.1](https://github.com/httpland/http-utils/compare/1.2.0...1.3.0-beta.1) (2023-05-09)


### Features

* mark as deprecated ([ab24b7d](https://github.com/httpland/http-utils/commit/ab24b7d2177aba3979a802fa02707a4a2a4ede9b))

# [1.2.0](https://github.com/httpland/http-utils/compare/1.1.0...1.2.0) (2023-04-29)


### Features

* **list:** add parser for list-based fields ([47f69bd](https://github.com/httpland/http-utils/commit/47f69bda005ef4adfabf34ad02e94e35cb51e06a))
* **token:** export related types ([35e0839](https://github.com/httpland/http-utils/commit/35e08398e1ac369f1c20f5365fe5668b1840f0ba))

# [1.2.0-beta.1](https://github.com/httpland/http-utils/compare/1.1.0...1.2.0-beta.1) (2023-04-29)


### Features

* **list:** add parser for list-based fields ([47f69bd](https://github.com/httpland/http-utils/commit/47f69bda005ef4adfabf34ad02e94e35cb51e06a))
* **token:** export related types ([35e0839](https://github.com/httpland/http-utils/commit/35e08398e1ac369f1c20f5365fe5668b1840f0ba))

# [1.1.0](https://github.com/httpland/http-utils/compare/1.0.0...1.1.0) (2023-04-29)


### Features

* **quoted_string:** add `isQdtext` function ([dfcd316](https://github.com/httpland/http-utils/commit/dfcd3167b159f5515aed07adf97fa2a48239ce48))
* **quoted_string:** add `isQuotedString` function ([ca5dee2](https://github.com/httpland/http-utils/commit/ca5dee2bdd585e5731c40980e3074e0ffd50a0fd))
* **quoted_string:** add `isQuotedString` function ([3c2f5a9](https://github.com/httpland/http-utils/commit/3c2f5a9b5dda5d6042f7db51620054991f824bd6))
* **token:** add http token utilities ([c11449f](https://github.com/httpland/http-utils/commit/c11449f772bd8e15abf51bb1ef68ff2e595ccb88))
* **token:** improve type inference ([a2b87b9](https://github.com/httpland/http-utils/commit/a2b87b9a4759121204c65878cd152265b2c9769b))

# [1.1.0-beta.1](https://github.com/httpland/http-utils/compare/1.0.0...1.1.0-beta.1) (2023-04-29)


### Features

* **quoted_string:** add `isQdtext` function ([dfcd316](https://github.com/httpland/http-utils/commit/dfcd3167b159f5515aed07adf97fa2a48239ce48))
* **quoted_string:** add `isQuotedString` function ([ca5dee2](https://github.com/httpland/http-utils/commit/ca5dee2bdd585e5731c40980e3074e0ffd50a0fd))
* **quoted_string:** add `isQuotedString` function ([3c2f5a9](https://github.com/httpland/http-utils/commit/3c2f5a9b5dda5d6042f7db51620054991f824bd6))
* **token:** add http token utilities ([c11449f](https://github.com/httpland/http-utils/commit/c11449f772bd8e15abf51bb1ef68ff2e595ccb88))
* **token:** improve type inference ([a2b87b9](https://github.com/httpland/http-utils/commit/a2b87b9a4759121204c65878cd152265b2c9769b))

# 1.0.0 (2023-04-02)


### Features

* **handler:** add types for http handler ([a0f3785](https://github.com/httpland/http-utils/commit/a0f3785df6686cd9f6e201bffb9d5befbcf1155e))
* **handler:** remove handler module ([44c719f](https://github.com/httpland/http-utils/commit/44c719fb3a3da3ce93698f86fa30c24948d94282))
* **handlers:** rename types ([cdf2721](https://github.com/httpland/http-utils/commit/cdf27215e4a33289b744903ec1f4afa4ff3f28cb))
* **header:** add caching header enum ([1844621](https://github.com/httpland/http-utils/commit/1844621b43293e362c41af24ef2f57dcf24eac4b))
* **header:** add conditional header fields ([863f4cb](https://github.com/httpland/http-utils/commit/863f4cb905e0b981de616acc58b54be0512f437c))
* **header:** add content negotiation header fields ([cc6c3d0](https://github.com/httpland/http-utils/commit/cc6c3d0eb0c3605d9b118e703fdda86c93b9237e))
* **header:** add fileld to representation header ([58965b2](https://github.com/httpland/http-utils/commit/58965b247926983b1e81462adcdb7eb69d2cc13e))
* **header:** add filtering by header field name function ([84c402c](https://github.com/httpland/http-utils/commit/84c402c5ab224f12bbd42df662c5552628d26507))
* **header:** add message forwarding header fields ([f2a439f](https://github.com/httpland/http-utils/commit/f2a439fcfa629b11813d6e66d02482c3f53e9cb8))
* **header:** add message metadata header fields ([75bb6ee](https://github.com/httpland/http-utils/commit/75bb6ee636af69c3814a79c19ac9a86ae781e498))
* **header:** add range requests header fields ([7c00c47](https://github.com/httpland/http-utils/commit/7c00c47fec3e0edd9cbf794b8ef602a0298d3eca))

header.ts
// Copyright 2023-latest the httpland authors. All rights reserved. MIT license.
// This module is browser compatible.

/** HTTP Message Metadata header fields.
 * @see [RFC 9110, 6.6. Message Metadata](https://www.rfc-editor.org/rfc/rfc9110.html#section-6.6)
 *
 * @example
 * ```ts
 * import { MessageMetadataHeader } from "https://deno.land/x/http_utils@$VERSION/header.ts";
 * import { assertEquals } from "https://deno.land/std/testing/asserts.ts";
 *
 * assertEquals(MessageMetadataHeader.Date, "date");
 * ```
 */
export enum MessageMetadataHeader {
  /**
   * @see [RFC 9110, 6.6.1. Date](https://www.rfc-editor.org/rfc/rfc9110.html#section-6.6.1)
   */
  Date = "date",

  /**
   * @see [RFC 9110, 6.6.2. Trailer](https://www.rfc-editor.org/rfc/rfc9110.html#section-6.6.2)
   */
  Trailer = "trailer",
}

/** HTTP Message Forwarding header fields.
 * @see [RFC 9110, 7.6. Message Forwarding](https://www.rfc-editor.org/rfc/rfc9110.html#section-7.6)
 *
 * @example
 * ```ts
 * import { MessageForwardingHeader } from "https://deno.land/x/http_utils@$VERSION/header.ts";
 * import { assertEquals } from "https://deno.land/std/testing/asserts.ts";
 *
 * assertEquals(MessageForwardingHeader.Via, "via");
 * ```
 */
export enum MessageForwardingHeader {
  /**
   * @see [RFC 9110, 7.6.1. Connection](https://www.rfc-editor.org/rfc/rfc9110.html#section-7.6.1)
   */
  Connection = "connection",

  /**
   * @see [RFC 9110, 7.6.2. Max-Forwards](https://www.rfc-editor.org/rfc/rfc9110.html#section-7.6.2)
   */
  MaxForwards = "max-forwards",

  /**
   * @see [RFC 9110, 7.6.3. Via](https://www.rfc-editor.org/rfc/rfc9110.html#section-7.6.3)
   */
  Via = "via",
}

/** HTTP representation data and metadata header fields.
 * @see [RFC 9110, 8. Representations](https://www.rfc-editor.org/rfc/rfc9110.html#section-8)
 *
 * @example
 * ```ts
 * import { RepresentationHeader } from "https://deno.land/x/http_utils@$VERSION/header.ts";
 * import { assertEquals } from "https://deno.land/std/testing/asserts.ts";
 *
 * assertEquals(RepresentationHeader.ContentType, "content-type");
 * ```
 */
export enum RepresentationHeader {
  /**
   * @see [RFC 9110, 8.3. Content-Type](https://www.rfc-editor.org/rfc/rfc9110.html#section-8.3)
   */
  ContentType = "content-type",

  /**
   * @see [RFC 9110, 8.4. Content-Encoding](https://www.rfc-editor.org/rfc/rfc9110.html#section-8.4)
   */
  ContentEncoding = "content-encoding",

  /**
   * @see [RFC 9110, 8.5. Content-Language](https://www.rfc-editor.org/rfc/rfc9110.html#section-8.5)
   */
  ContentLanguage = "content-language",

  /**
   * @see [RFC 9110, 8.6. Content-Length](https://www.rfc-editor.org/rfc/rfc9110.html#section-8.6)
   */
  ContentLength = "content-length",

  /**
   * @see [RFC 9110, 8.7. Content-Location](https://www.rfc-editor.org/rfc/rfc9110.html#section-8.7)
   */
  ContentLocation = "content-location",

  /**
   * @see [RFC 9110, 8.8.2. Last-Modified](https://www.rfc-editor.org/rfc/rfc9110.html#section-8.8.2)
   */
  LastModified = "last-modified",

  /**
   * @see [RFC 9110, 8.8.3. ETag](https://www.rfc-editor.org/rfc/rfc9110.html#section-8.8.3)
   */
  ETag = "etag",
}

/** HTTP Authentication header fields.
 * @see [RFC 9110, 11. HTTP Authentication](https://www.rfc-editor.org/rfc/rfc9110#section-11)
 */
export enum AuthenticationHeader {
  /**
   * @see [RFC 9110, 11.6.1. WWW-Authenticate](https://www.rfc-editor.org/rfc/rfc9110#section-11.6.1)
   */
  WWWAuthenticate = "www-authenticate",

  /**
   * @see [RFC 9110, 11.6.2. Authorization](https://www.rfc-editor.org/rfc/rfc9110#section-11.6.2)
   */
  Authorization = "authorization",

  /**
   * @see [RFC 9110, 11.6.3. Authentication-Info](https://www.rfc-editor.org/rfc/rfc9110#section-11.6.3)
   */
  AuthenticationInfo = "authentication-info",

  /**
   * @see [RFC 9110, 11.7.1. Proxy-Authenticate](https://www.rfc-editor.org/rfc/rfc9110#section-11.7.1)
   */
  ProxyAuthenticate = "proxy-authenticate",

  /**
   * @see [RFC 9110, 11.7.2. Proxy-Authorization](https://www.rfc-editor.org/rfc/rfc9110#section-11.7.2)
   */
  ProxyAuthorization = "proxy-authorization",

  /**
   * @see [RFC 9110, 11.7.3. Proxy-Authentication-Info](https://www.rfc-editor.org/rfc/rfc9110#section-11.7.3)
   */
  ProxyAuthenticationInfo = "proxy-authentication-info",
}

/** HTTP content negotiation header fields.
 * @see [RFC 9110, 12. Content Negotiation](https://www.rfc-editor.org/rfc/rfc9110#section-12)
 *
 * @example
 * ```ts
 * import { ContentNegotiationHeader } from "https://deno.land/x/http_utils@$VERSION/header.ts";
 * import { assertEquals } from "https://deno.land/std/testing/asserts.ts";
 *
 * assertEquals(ContentNegotiationHeader.Accept, "accept");
 * ```
 */
export enum ContentNegotiationHeader {
  /**
   * @see [RFC 9110, 12.5.1. Accept](https://www.rfc-editor.org/rfc/rfc9110#section-12.5.1)
   */
  Accept = "accept",
  /**
   * @see [RFC 9110, 12.5.2. Accept-Charset](https://www.rfc-editor.org/rfc/rfc9110#section-12.5.2)
   */
  AcceptCharset = "accept-charset",
  /**
   * @see [RFC 9110, 12.5.3. Accept-Encoding](https://www.rfc-editor.org/rfc/rfc9110#section-12.5.3)
   */
  AcceptEncoding = "accept-encoding",
  /**
   * @see [RFC 9110, 12.5.4. Accept-Language](https://www.rfc-editor.org/rfc/rfc9110#section-12.5.4)
   */
  AcceptLanguage = "accept-language",
  /**
   * @see [RFC 9110, 12.5.5. Vary](https://www.rfc-editor.org/rfc/rfc9110#section-12.5.5)
   */
  Vary = "vary",
}

/** HTTP conditional requests header fields.
 * @see [RFC 9110, 13. Conditional Requests](https://www.rfc-editor.org/rfc/rfc9110#section-13)
 *
 * @example
 * ```ts
 * import { ConditionalHeader } from "https://deno.land/x/http_utils@$VERSION/header.ts";
 * import { assertEquals } from "https://deno.land/std/testing/asserts.ts";
 *
 * assertEquals(ConditionalHeader.IfNoneMatch, "if-none-match");
 * ```
 */
export enum ConditionalHeader {
  /**
   * @see [RFC 9110, 13.1.1. If-Match](https://www.rfc-editor.org/rfc/rfc9110#section-13.1.1)
   */
  IfMatch = "if-match",
  /**
   * @see [RFC 9110, 13.1.2. If-None-Match](https://www.rfc-editor.org/rfc/rfc9110#section-13.1.2)
   */
  IfNoneMatch = "if-none-match",
  /**
   * @see [RFC 9110, 13.1.3. If-Modified-Since](https://www.rfc-editor.org/rfc/rfc9110#section-13.1.3)
   */
  IfModifiedSince = "if-modified-since",
  /**
   * @see [RFC 9110, 13.1.4. If-Unmodified-Since](https://www.rfc-editor.org/rfc/rfc9110#section-13.1.4)
   */
  IfUnmodifiedSince = "if-unmodified-since",
  /**
   * @see [RFC 9110, 13.1.5. If-Range](https://www.rfc-editor.org/rfc/rfc9110#section-13.1.5)
   */
  IfRange = "if-range",
}

/** HTTP range requests header fields.
 * @see [RFC 9110, 14. Range Requests](https://www.rfc-editor.org/rfc/rfc9110#section-14)
 *
 * @example
 * ```ts
 * import { RangeHeader } from "https://deno.land/x/http_utils@$VERSION/header.ts";
 * import { assertEquals } from "https://deno.land/std/testing/asserts.ts";
 *
 * assertEquals(RangeHeader.Range, "range");
 * ```
 */
export enum RangeHeader {
  /**
   * @see [RFC 9110, 14.2. Range](https://www.rfc-editor.org/rfc/rfc9110#section-14.2)
   */
  Range = "range",
  /**
   * @see [RFC 9110, 14.3. Accept-Ranges](https://www.rfc-editor.org/rfc/rfc9110#section-14.3)
   */
  AcceptRanges = "accept-ranges",
  /**
   * @see [RFC 9110, 14.4. Content-Range](https://www.rfc-editor.org/rfc/rfc9110#section-14.4)
   */
  ContentRange = "content-range",
}

/** HTTP Caching header fields.
 * @see [RFC 9111, HTTP Caching](https://www.rfc-editor.org/rfc/rfc9111)
 *
 * @example
 * ```ts
 * import { CachingHeader } from "https://deno.land/x/http_utils@$VERSION/header.ts";
 * import { assertEquals } from "https://deno.land/std/testing/asserts.ts";
 *
 * assertEquals(CachingHeader.CacheControl, "cache-control");
 * ```
 */
export enum CachingHeader {
  /**
   * @see [RFC 9111, 5.1. Age](https://www.rfc-editor.org/rfc/rfc9111#section-5.1)
   */
  Age = "age",

  /**
   * @see [RFC 9111, 5.2. Cache-Control](https://www.rfc-editor.org/rfc/rfc9111#section-5.2)
   */
  CacheControl = "cache-control",

  /**
   * @see [RFC 9111, 5.3. Expires](https://www.rfc-editor.org/rfc/rfc9111#section-5.3)
   */
  Expires = "expires",
}

const MESSAGE_METADATA_HEADERS = [
  MessageMetadataHeader.Date,
  MessageMetadataHeader.Trailer,
];

/** Whether the input is {@link MessageMetadataHeader} or not.
 *
 * @example
 * ```ts
 * import { isMessageMetadataHeader } from "https://deno.land/x/http_utils@$VERSION/header.ts";
 * import { assert } from "https://deno.land/std/testing/asserts.ts";
 *
 * assert(isMessageMetadataHeader("date"));
 * assert(!isMessageMetadataHeader("<others>"));
 * ```
 */
export function isMessageMetadataHeader(
  input: string,
): input is `${MessageMetadataHeader}` {
  return (MESSAGE_METADATA_HEADERS as string[]).includes(input);
}

const MESSAGE_FORWARDING_HEADERS = [
  MessageForwardingHeader.Connection,
  MessageForwardingHeader.MaxForwards,
  MessageForwardingHeader.Via,
];

/** Whether the input is {@link MessageForwardingHeader} or not.
 *
 * @example
 * ```ts
 * import { isMessageForwardingHeader } from "https://deno.land/x/http_utils@$VERSION/header.ts";
 * import { assert } from "https://deno.land/std/testing/asserts.ts";
 *
 * assert(isMessageForwardingHeader("connection"));
 * assert(!isMessageForwardingHeader("<others>"));
 * ```
 */
export function isMessageForwardingHeader(
  input: string,
): input is `${MessageForwardingHeader}` {
  return (MESSAGE_FORWARDING_HEADERS as string[]).includes(input);
}

const REPRESENTATION_HEADERS = [
  RepresentationHeader.ContentEncoding,
  RepresentationHeader.ContentLanguage,
  RepresentationHeader.ContentLength,
  RepresentationHeader.ContentLocation,
  RepresentationHeader.ContentType,
  RepresentationHeader.ETag,
  RepresentationHeader.LastModified,
];

/** Whether the input is {@link RepresentationHeader} or not.
 *
 * @example
 * ```ts
 * import { isRepresentationHeader } from "https://deno.land/x/http_utils@$VERSION/header.ts";
 * import { assert } from "https://deno.land/std/testing/asserts.ts";
 *
 * assert(isRepresentationHeader("content-type"));
 * assert(!isRepresentationHeader("<others>"));
 * ```
 */
export function isRepresentationHeader(
  input: string,
): input is `${RepresentationHeader}` {
  return (REPRESENTATION_HEADERS as string[]).includes(input);
}

const AUTHENTICATION_HEADERS = [
  AuthenticationHeader.AuthenticationInfo,
  AuthenticationHeader.Authorization,
  AuthenticationHeader.ProxyAuthenticate,
  AuthenticationHeader.ProxyAuthenticationInfo,
  AuthenticationHeader.ProxyAuthorization,
  AuthenticationHeader.WWWAuthenticate,
];

/** Whether the input is {@link AuthenticationHeader} or not.
 *
 * @example
 * ```ts
 * import { isAuthenticationHeader } from "https://deno.land/x/http_utils@$VERSION/header.ts";
 * import { assert } from "https://deno.land/std/testing/asserts.ts";
 *
 * assert(isAuthenticationHeader("authorization"));
 * assert(!isAuthenticationHeader("<others>"));
 * ```
 */
export function isAuthenticationHeader(
  input: string,
): input is `${AuthenticationHeader}` {
  return (AUTHENTICATION_HEADERS as string[]).includes(input);
}

const CONTENT_NEGOTIATION_HEADERS = [
  ContentNegotiationHeader.Accept,
  ContentNegotiationHeader.AcceptCharset,
  ContentNegotiationHeader.AcceptEncoding,
  ContentNegotiationHeader.AcceptLanguage,
  ContentNegotiationHeader.Vary,
];

/** Whether the input is {@link ContentNegotiationHeader} or not.
 *
 * @example
 * ```ts
 * import { isContentNegotiationHeader } from "https://deno.land/x/http_utils@$VERSION/header.ts";
 * import { assert } from "https://deno.land/std/testing/asserts.ts";
 *
 * assert(isContentNegotiationHeader("accept"));
 * assert(!isContentNegotiationHeader("<others>"));
 * ```
 */
export function isContentNegotiationHeader(
  input: string,
): input is `${ContentNegotiationHeader}` {
  return (CONTENT_NEGOTIATION_HEADERS as string[]).includes(input);
}

const CONDITIONAL_HEADERS = [
  ConditionalHeader.IfMatch,
  ConditionalHeader.IfModifiedSince,
  ConditionalHeader.IfNoneMatch,
  ConditionalHeader.IfRange,
  ConditionalHeader.IfUnmodifiedSince,
];

/** Whether the input is {@link ConditionalHeader} or not.
 *
 * @example
 * ```ts
 * import { isConditionalHeader } from "https://deno.land/x/http_utils@$VERSION/header.ts";
 * import { assert } from "https://deno.land/std/testing/asserts.ts";
 *
 * assert(isConditionalHeader("if-match"));
 * assert(!isConditionalHeader("<others>"));
 * ```
 */
export function isConditionalHeader(
  input: string,
): input is `${ConditionalHeader}` {
  return (CONDITIONAL_HEADERS as string[]).includes(input);
}

const RANGE_HEADERS = [
  RangeHeader.AcceptRanges,
  RangeHeader.ContentRange,
  RangeHeader.Range,
];

/** Whether the input is {@link RangeHeader} or not.
 *
 * @example
 * ```ts
 * import { isRangeHeader } from "https://deno.land/x/http_utils@$VERSION/header.ts";
 * import { assert } from "https://deno.land/std/testing/asserts.ts";
 *
 * assert(isRangeHeader("range"));
 * assert(!isRangeHeader("<others>"));
 * ```
 */
export function isRangeHeader(input: string): input is `${RangeHeader}` {
  return (RANGE_HEADERS as string[]).includes(input);
}

const CACHING_HEADERS = [
  CachingHeader.Age,
  CachingHeader.CacheControl,
  CachingHeader.Expires,
];

/** Whether the input is {@link CachingHeader} or not.
 *
 * @example
 * ```ts
 * import { isCachingHeader } from "https://deno.land/x/http_utils@$VERSION/header.ts";
 * import { assert } from "https://deno.land/std/testing/asserts.ts";
 *
 * assert(isCachingHeader("age"));
 * assert(!isCachingHeader("<others>"));
 * ```
 */
export function isCachingHeader(input: string): input is `${CachingHeader}` {
  return (CACHING_HEADERS as string[]).includes(input);
}
